### ISSUE TITLE
Silent `Realm.Write` failure when Realm instance thread is different

### ISSUE THREAD

**Issue by sushihangover** [realm/realm-dotnet#1261]
#### Goal

When calling `Realm.Write` within a `Timer` event hander/descendent, the `Realm.Write` does not error and thus the data is not persisted and a data integrity issue develops.

#### Expected Results

Using a `Realm.Write` on a non-originating thread will throw an exception.

#### Actual Results

Full stack trace with exception:  None, that is the problem.. ;-) Nothing reported in Logcat either

#### Steps & Code to Reproduce

I was debugging some client code, and they have a `System.Timer.Timer` that causes updates to a `Realm` instance based on some user interaction on the UI looper. Since the timer "callback" is not guaranteed to be on the originating thread, `Realm.Write` should produce an exception but is silently failing and thus the client's Realm data is left in an unknown state as some `Realm.Write` are working and others not... This condition should throw an exception to prevent data integrity issues.

#### Version of Realm and tooling

`Realm.Database" version="1.0.4" targetFramework="monoandroid71"
`

####Which operating system version and device: ?

Only Android was tested 

#### Code Sample

The following will repo the issue:

```csharp

	public class ARealmObject : RealmObject
	{
		public int key { get; set; }
	}

	[Activity(Label = "RealmTimerSilentFailure", MainLauncher = true, Icon = "@mipmap/icon")]
	public class MainActivity : Activity
	{
		Realm realm;
		Timer timer;
		bool toggle;
		int recordCount;

		protected override void OnCreate(Bundle savedInstanceState)
		{
			base.OnCreate(savedInstanceState);
			SetContentView(Resource.Layout.Main);
			Button button = FindViewById<Button>(Resource.Id.myButton);

			realm = Realm.GetInstance("10");
			timer = new Timer(250);
			timer.Elapsed += Timer_Elapsed;

			button.Click += delegate
			{
				realm.Write(() => { realm.RemoveAll(); });
				recordCount = 0;
				timer.Start();
			};
		}

		void Timer_Elapsed(object sender, ElapsedEventArgs e)
		{
			Log.Debug("REALM", "Timer_Elapsed");
			recordCount++;
			toggle = !toggle;
			if (recordCount == 10)
			{
				timer.Stop();
				// get new realm in case we are on a different thread...
				var newRealm = Realm.GetInstance(realm.Config);
				Log.Debug("REALM", $"Count {newRealm.All<ARealmObject>().Count()}");
				System.Diagnostics.Debug.Assert(newRealm.All<ARealmObject>().Count() == 10);
				return;
			}

			if (toggle)
			{
				realm.Write(() => // silent failure (appears "random" to the client but of course is a thread issue)
				{
					Log.Debug("REALM", "Write");
					var obj = new ARealmObject { key = recordCount };
					realm.Add(obj);
				});
			}
			else
			{
				var newRealm = Realm.GetInstance(realm.Config); // Works fine of course...
				newRealm.Write(() =>
				{
					Log.Debug("REALM", "Write on new instance");
					var obj = new ARealmObject { key = recordCount };
					newRealm.Add(obj);
				});
			}
			timer.Start();
		}
	}
```

**Comment by sushihangover** [comment]
Note: Using `realm.WriteAsync` to try to avoid the issue instead of obtaining a "new" Realm instance first, will cause a infinite `Task` continuation hang within the UI looper.

**Comment by sushihangover** [comment]
Version 1.1.0 produces the same silent failures:

`  <package id="Realm.Database" version="1.1.0" targetFramework="monoandroid71" />
`

**Comment by AndyDentFree** [comment]
I think this overlaps with the threading behaviour I'm investigating in #1204 so will lump it in with that.

**Comment by sushihangover** [comment]
FYI: I have another client who is getting "silent" failures on `Realm.WriteAsync`. These are on iOS 10 using Realm.Database 1.1.0 (simulator and physical).

There are Realm Assertions in the iOS log:

```
Assertion failed: (ret >= 0), function listen, file /Users/realm/workspace/realm_realm-dotnet_PR-1259-S2QELV65P7BDGRCDIBL2GFPGLFT2KXZLPVJ5URX7TYYZ6DVN3OAQ/wrappers/src/object-store/src/impl/apple/external_commit_helper.cpp, line 203.
2017-03-07 08:17:44.915 GeneticAnalysisWizard.iOS[18469:42726523] Xamarin.iOS: Processing: 'exit process'


```

**Comment by nirinchev** [comment]
I may be misunderstanding something, but from the provided code sample, it seems like there's no exception handling. Since the timer executes on a background thread, the write will throw an unobserved exception - it can't be propagated to the main thread and this is the way any .NET code will behave. If you replace the `Timer_Elapsed` implementation with `throw new Exception("Test");`, you'll get the same results. So in that aspect, the exception being thrown by realm is not "silent", it's just that no one is observing it and since it's on a background thread, it doesn't crash the app.

Since to me, this seems like the expected .NET behavior, can you elaborate on what you would like to see improved?

**Comment by sushihangover** [comment]
From the client's perspective, the issue came from the fact that the `realm` variable became a captured variable on a different thread and eventually gets passed into a method that updates a Realm instance. They did not intentionally plan that, the logic path/callstack was complex, but it did happen and it appears the database was updated and the transaction was completed and now the database is in a fubar™ state.

`System.Threading.Timer` of course will propagate exceptions and cause a process termination, while usage of `System.Timers.Timer` eats them as is their usage/case and as a result any unhandled exception handler never gets any called for the "Realm accessed from incorrect thread" exception. They were not interested in converting to `System.Threading.Timer` as this is in shared libraries so I wrote them a wrapper and use the `SynchronizationContext` of the `realm` instance to propagate any unhandled exceptions (thus their crash reporter grabs it since it is registered on the main thread). Last I heard, this has surfaced a number of cross-thread "silent" exceptions for them...

So in the end, they "assumed" Realm would always throw "somehow", even if it is a "native" exception/crash when it comes to data integrity. I think most devs would also expect this also. But that really comes down to how Realm views data integrity vs. particular language constructs.

**Comment by nirinchev** [comment]
The thing is, for us that's not a data integrity issue - accessing the realm from an incorrect thread is a common error, that we throw managed exception for. We don't want to crash the app in that case. The fact that they didn't handle the exception is a programmer error that I don't see an easy solution for. The same would happen if they started a Task.Run and an exception occurred there.

@fealebenpae @kristiandupont what's your take on that?

**Comment by kristiandupont** [comment]
I agree with you @nirinchev. Realm considers data integrity at level of transactions -- we provide the ACID guarantee. On levels higher than that, the user has to design for it. 
I can see that this situation is unfortunate but as it's aligned with standard .NET behavior I don't think it's an issue with Realm.

**Comment by sushihangover** [comment]
I have to disagree, trying to `Write` (`.BeginTransaction`) to a Realm via `Task.Run` used with a cross-thread captured Realm instance will throw and end the process (System.Exception: Realm accessed from incorrect thread), unless it is handled via `AppDomain.UnhandledException`.

If you are running in .Net runtime legacy mode, which I do not know if it is even a supported environment within Xamarin,  things would be different of course...

**Comment by nirinchev** [comment]
That is unexpected for me behavior and I did a very quick test with a XF app. In the main page, after `InitializeComponent`, I put

```csharp
var realm = Realm.GetInstance();
Task.Run(() => 
{
    var test = realm.All<Foo>().Count();
});
```

and when I ran the app, it worked just fine. Obviously, if I wrap the `realm.All` call in a try-catch block, I hit a breakpoint in the catch clause and the exception is indeed realm telling me it's being used by an incorrect thread.

Similarly, if I replace the `Timer_Elapsed` implementation from your customer's code with `throw new Exception()`, their app will not crash, the reason being managed exceptions on a background thread will not bring the whole process down. I have a feeling (which I haven't verified) that if we throw a native exception, that will crash the app, but that's definitely not an approach I'd want to take.

In my experience, native crashes have been very frustrating and very annoying to track down and would rather we avoid introducing them in users' applications. Do you feel that's preferable to the managed exceptions approach we're taking?

**Comment by sushihangover** [comment]
Without a doubt I prefer managed exceptions to native one and when used with `EndInvoke` via the `Task` thread pool they will *always* propagate to the `AppDomain` if unhandled and thus end the process if not handled there (assuming you are not running in legacy mode). Thus apps that are using crash reporters log that crash, report them on next run, devs get the crash report, fix their code, release updated app, raise, repeat... 90% of .Net devs *assume* threading will not ever be an issue due to async/await and all the other protection the .Net framework provides, but with Realm, they are now living in a thread context based world and for most of those LoB developers, this is the first time something is thread sensitive other than UI updates and even then Microsoft hides almost all of that behind sync contexts. But coding problems can only be fixed if the developer know about them and IMHO an API can assist the developer but ensuring that an exception is raised "properly" by the API if they are not handle in the user's code.

That said there are a number of special cases in the .Net framework that even Microsoft knows is a total kludge, one being the `System.Timers` namespace. `System.Timers.Timer` breaks the standard flow/processing of handlers, i.e. it eats exceptions. One of the "original" reasons to not "fix" it and others like it in the 2.0 runtime and 2.0+ framework release is that it is used as a timer within server services and having your entire service died on every spun-off handler exception would be wrong and too big of a breaking change in 2.0, I sure they much more intelligent devs than myself working on it and their reasoning was sound at the time. Years later, it is so widely used by .Net devs because it is so dang simple to code. Now `System.Threading.Timer` which "under the hood" uses the Task-based thread pool and thus will propagate exceptions "properly" *should* be used in the modern async/await world but its API is confusing (the no.1 excuse I hear, BS, but people use what they know and understand...) Forward to today, items like `System.Timers.Timer` are flagged as to never be supported in .Net Core, .Net Standard, etc...due to these inherent problems but we still have to live with them for many years to come.

I know of three Realm users that have had to deal with corrupted Realm instances in the field due to `System.Timers.Timer` *style* issues with threads that all used `Realm.Write`. As someone who understand the reason behind the issue, I could say you got with you deserve for using it that way, but I also greatly feel their pain as they all *assumed* the Realm API would protect them and now they have mad customers, have to figure out what data is valid and can be saved in a forced Realm migration, and they have to figure out how to fix the root problem **if** they even know what the problem is...

My simple solution for one that had contracted me to help was to add an Realm class extension method (`.SafeWrite`) that wraps the `Realm.Write` in a try/catch and thus for any exception that the user's Realm.Write lambda is not handling directly, ends up in the extension's catch and using exception filters on *only* Realm exceptions, *re-throw* them in a 2.0+ style via `ThreadPool.QueueUserWorkItem` (or a `Task.Run`, etc.. ). The reasoning of filtering only for Realm is we are in a *Realm.Write* that is automatically handling the transaction and any non-handled Realm failure is a serious issue as we are dealing with the integrity of the persisted user's data and **have** to be notified of any failures related to persisting this data. Thus a simple search and replace of `Write` to `SafeWrite` gets them to a point that non-handled "hidden" Realm exceptions on cross-threaded event and timer handlers will be raised all the way to the AppDomain only if no other users' try/catch is in the call stack, this is in the *exact* same way as if they were using a `Realm.Write` via a Tasks or Action or using the wrong Realm instance within a `WriteAsync` lambda, etc... 

Thus the `Write` forces an exception to be raised (even if on a different thread), they are notified of the failure type (Realm thread exception, Invalid list update, etc.) and they *feel* the Realm API is protecting them again because it is notifying them of where the problem is, guiding them to fix their app, and the user's code becomes better due.
